src/types/decimal.rs

```diff
@@ -167,6 +168,63 @@ impl Decimal {
         scaled_coefficient *= BigUint::from(10u64).pow(exponent_delta as u32);
         UInt::BigUInt(scaled_coefficient).cmp(d2.coefficient.magnitude())
     }
+
+    /// Extract the integer and fractional parts of this as a pair of [`num_bigint::BigInt`]s
+    ///
+    /// ```ignore
+    /// # use num_bigint::BigInt;
+    /// # use ion_rs::Decimal;
+    /// let d1: Decimal = Decimal::new(123456789, -2);
+    /// let (int_part, frac_part) = d1.into_parts();
+    /// assert_eq!(int_part, BigInt::from(1234567u64));
+    /// assert_eq!(frac_part, BigInt::from(89));
+    /// ```
+    pub(crate) fn into_parts(self) -> (BigInt, BigInt) {
+        let magnitude: BigInt = self.coefficient.try_into().unwrap();
+        if self.exponent.is_zero() {
+            (magnitude, BigInt::zero())
+        } else if self.exponent.is_negative() {
+            let divisor = BigInt::from(10u64).pow((-self.exponent) as u32);
+            magnitude.div_rem(&divisor)
+        } else {
+            let multiplicand = BigInt::from(10u64).pow(self.exponent as u32);
+            (magnitude * multiplicand, BigInt::zero())
+        }
+    }
+
+    /// Returns ([`num_bigint::BigInt`], [`f64`]) representing the differences
+    /// between integer and fractional components of d1 & d2 respectively
+    ///
+    /// ```ignore
+    /// # use num_bigint::BigInt;
+    /// # use ion_rs::Decimal;
+    /// let (diff_integer, diff_fractional) = Decimal::difference_by_parts(Decimal::new(123, 0), Decimal::new(12, 1));
+    /// assert_eq!(diff_integer, BigInt::from(3));
+    /// assert_eq!(diff_fractional, 0.0);
+    ///
+    /// let (diff_integer, diff_fractional) = Decimal::difference_by_parts(Decimal::new(12345, -2), Decimal::new(123, 0));
+    /// assert_eq!(diff_integer, BigInt::from(0));
+    /// assert_eq!(diff_fractional, 0.45);
+    /// ```
+    pub(crate) fn difference_by_parts(d1: Decimal, d2: Decimal) -> (BigInt, f64) {
+        let (d1_int, d1_frac) = d1.into_parts();
+        let (d2_int, d2_frac) = d2.into_parts();
+
+        // turn a fractional part (e.g., `123`) into an f64 less than `0` (e.g., `0.123`)
+        fn to_fract(frac: BigInt) -> f64 {
+            if frac.is_zero() {
+                0.0
+            } else {
+                let frac_oom = max(1, 1 + frac.abs().to_u128().unwrap().ilog10());
+                frac.to_f64().unwrap() / 10f64.powi(frac_oom as i32)
+            }
```

Comment:
Could you add a comment here explaining the logic? I'm also unsure what `_oom` means here, my brain can't get away from `OutOfMemory`.

Reply:
will do. Order Of Magnitude.